alibi_detect/cd/keops/mmd.py

```diff
@@ -0,0 +1,179 @@
+            data will be preprocessed at prediction time. Only applies if `x_ref_preprocessed=False`.
+        update_x_ref
+            Reference data can optionally be updated to the last n instances seen by the detector
+            or via reservoir sampling with size n. For the former, the parameter equals {'last': n} while
+            for reservoir sampling {'reservoir_sampling': n} is passed.
+        preprocess_fn
+            Function to preprocess the data before computing the data drift metrics.
+        kernel
+            Kernel used for the MMD computation, defaults to Gaussian RBF kernel.
+        sigma
+            Optionally set the GaussianRBF kernel bandwidth. Can also pass multiple bandwidth values as an array.
+            The kernel evaluation is then averaged over those bandwidths.
+        configure_kernel_from_x_ref
+            Whether to already configure the kernel bandwidth from the reference data.
+        n_permutations
+            Number of permutations used in the permutation test.
+        batch_size_permutations
+            KeOps computes the n_permutations of the MMD^2 statistics in chunks of batch_size_permutations.
+        device
+            Device type used. The default None tries to use the GPU and falls back on CPU if needed.
+            Can be specified by passing either 'cuda', 'gpu' or 'cpu'.
+        input_shape
+            Shape of input data.
+        data_type
+            Optionally specify the data type (tabular, image or time-series). Added to metadata.
+        """
+        super().__init__(
+            x_ref=x_ref,
+            p_val=p_val,
+            x_ref_preprocessed=x_ref_preprocessed,
+            preprocess_at_init=preprocess_at_init,
+            update_x_ref=update_x_ref,
+            preprocess_fn=preprocess_fn,
+            sigma=sigma,
+            configure_kernel_from_x_ref=configure_kernel_from_x_ref,
+            n_permutations=n_permutations,
+            input_shape=input_shape,
+            data_type=data_type
+        )
+        self.meta.update({'backend': 'keops'})
+
+        # set device
+        self.device = get_device(device)
+
+        # initialize kernel
+        sigma = torch.from_numpy(sigma).to(self.device) if isinstance(sigma,  # type: ignore[assignment]
+                                                                      np.ndarray) else None
+        self.kernel = kernel(sigma).to(self.device) if kernel == GaussianRBF else kernel
+
+        # set the correct MMD^2 function based on the batch size for the permutations
+        self.batch_size = batch_size_permutations
+        self.n_batches = 1 + (n_permutations - 1) // batch_size_permutations
+
+        # infer the kernel bandwidth from the reference data
+        if self.infer_sigma or isinstance(sigma, torch.Tensor):
+            x = torch.from_numpy(self.x_ref).to(self.device)
+            _ = self.kernel(LazyTensor(x[:, None, :]), LazyTensor(x[None, :, :]), infer_sigma=self.infer_sigma)
```

Comment:
If `self.infer_sigma = False` and `sigma` is passed by user then isn't this call redundant?

Reply:
Yes good catch. Will change to:
```python
if isinstance(sigma, torch.Tensor):
   self.infer_sigma = False
elif self.infer_sigma:
   x = torch.from_numpy(self.x_ref).to(self.device)
    _ = self.kernel(LazyTensor(x[:, None, :]), LazyTensor(x[None, :, :]), infer_sigma=self.infer_sigma)
    self.infer_sigma = False
else:
   self.infer_sigma = True
```

Reply:
Should the same change be made for the `torch` and `tensorflow` ones (different PR)? I'm questioning this now as it's present in these two so perhaps there was some reason to do it this way, otherwise looks like a bug.

Reply:
I don't think so since there are 2 reasons to do this:
1. Infer a kernel bandwidth sigma.
2. Obtain kernel matrix `k_xx`.
These can be grouped for PyTorch/TensorFlow but not KeOps since there is no benefit from computing a symbolic `k_xx` in advance, so you only do it to infer sigma which as you noticed made the statement a bit redundant.